### DISCUSSION TITLE
Handling dynamically loaded libraries with go-appimagetool

### DISCUSSION THREAD

**Discussion by VioletGiraffe** [AppImage/AppImageKit#1297] [Q&A]
Hi! I have an issue with deploying libcamera into AppImage. It has some dynamic modules that are loaded at runtime (not listed by `ldd`). Libcamera looks for them in ` /usr/local/lib/aarch64-linux-gnu/libcamera/`. I have created a folder `AppDir/usr/local/lib/aarch64-linux-gnu/libcamera/` and copied the necessary stuff there before running `appimagetool`, but it's not enough. If, however, I create this system folder `/usr/local/lib/aarch64-linux-gnu/libcamera/` and copy the extra libraries there on the clean target system, then the AppImage works. Is there a way to fix this, patch absolute search paths inside binaries or something? On Windows you could hook the `CreateFile` function and replace its arguments, but I don't know if there's a similar facility on Linux.

P. S. I mentioned the Go `appimagetool` specifically because I need to get it working on ARM64.

**Comment by probonopd** [comment]
What happens if you export `LD_LIBRARY_PATH` and the run `appimagetool`?

```
export LD_LIBRARY_PATH=/usr/local/lib/aarch64-linux-gnu/libcamera/
appimagetool [-s] deploy (...)
```

**Reply by VioletGiraffe** [reply]
Thanks a lot for responding! Your suggestion seems to have made no difference, the AppDir does not have .so files from `/usr/local/lib/aarch64-linux-gnu/libcamera/` in my `AppDir/usr/local/lib/aarch64-linux-gnu/libcamera/`. And, of course, when this AppImage runs on the clean system, libcamera complains that it can't find its modules.

**Comment by probonopd** [comment]
Which application are you trying to make an AppImage of?

**Reply by VioletGiraffe** [reply]
It's my own app, but any of the official libcamera-apps should behave exactly the same if you try to deploy them into an AppImage. For example, `libcamera-jpeg` from the `libcamera-apps` apt package. I do use the `-s` flag.

**Reply by probonopd** [reply]
What error messages are you getting?

What does `ldd ./your/application | grep libcamera` say?

**Reply by VioletGiraffe** [reply]
The error when running the AppImage is this (there are no errors when assembling the AppImage): 
```
WARN IPAManager ipa_manager.cpp:154 No IPA found in '/usr/local/lib/aarch64-linux-gnu/libcamera
```
ldd does not pick up the IPA .so files as dependencies, that's exactly the problem. They are, apparently, discovered and loaded at runtime.,

**Reply by probonopd** [reply]
What does `ldd ./your/application | grep libcamera` say? Before you run `appimagetool` on it.

And what is an IPA?

**Reply by probonopd** [reply]
Is your application on GitHub?

**Reply by VioletGiraffe** [reply]
Sorry, it's not available, but any of the `libcamera-apps` should work exactly the same as a demonstrator. Here's the output:
```
ldd bin/MyTest
        linux-vdso.so.1 (0x0000007fb3cd8000)
        libudev.so.1 => /lib/aarch64-linux-gnu/libudev.so.1 (0x0000007fb3b20000)
        libcamera-base.so.0.1 => /usr/local/lib/aarch64-linux-gnu/libcamera-base.so.0.1 (0x0000007fb3a70000)
        libcamera.so.0.1 => /usr/local/lib/aarch64-linux-gnu/libcamera.so.0.1 (0x0000007fb3310000)
        libQt6Gui.so.6 => /lib/aarch64-linux-gnu/libQt6Gui.so.6 (0x0000007fb2c10000)
        libQt6Core.so.6 => /lib/aarch64-linux-gnu/libQt6Core.so.6 (0x0000007fb2720000)
        libstdc++.so.6 => /lib/aarch64-linux-gnu/libstdc++.so.6 (0x0000007fb2500000)
        libm.so.6 => /lib/aarch64-linux-gnu/libm.so.6 (0x0000007fb2460000)
        libgcc_s.so.1 => /lib/aarch64-linux-gnu/libgcc_s.so.1 (0x0000007fb2420000)
        libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000007fb2270000)
        /lib/ld-linux-aarch64.so.1 (0x0000007fb3c9b000)
        libdw.so.1 => /lib/aarch64-linux-gnu/libdw.so.1 (0x0000007fb21b0000)
        libpisp.so.1.0.0 => /usr/local/lib/aarch64-linux-gnu/libpisp.so.1.0.0 (0x0000007fb2080000)
        libcrypto.so.3 => /lib/aarch64-linux-gnu/libcrypto.so.3 (0x0000007fb1c20000)
        libyaml-0.so.2 => /lib/aarch64-linux-gnu/libyaml-0.so.2 (0x0000007fb1be0000)
        libEGL.so.1 => /lib/aarch64-linux-gnu/libEGL.so.1 (0x0000007fb1ba0000)
        libfontconfig.so.1 => /lib/aarch64-linux-gnu/libfontconfig.so.1 (0x0000007fb1b30000)
        libX11.so.6 => /lib/aarch64-linux-gnu/libX11.so.6 (0x0000007fb19d0000)
        libglib-2.0.so.0 => /lib/aarch64-linux-gnu/libglib-2.0.so.0 (0x0000007fb1870000)
        libQt6DBus.so.6 => /lib/aarch64-linux-gnu/libQt6DBus.so.6 (0x0000007fb17a0000)
        libxkbcommon.so.0 => /lib/aarch64-linux-gnu/libxkbcommon.so.0 (0x0000007fb1730000)
        libGLX.so.0 => /lib/aarch64-linux-gnu/libGLX.so.0 (0x0000007fb16d0000)
        libOpenGL.so.0 => /lib/aarch64-linux-gnu/libOpenGL.so.0 (0x0000007fb1650000)
        libpng16.so.16 => /lib/aarch64-linux-gnu/libpng16.so.16 (0x0000007fb15f0000)
        libharfbuzz.so.0 => /lib/aarch64-linux-gnu/libharfbuzz.so.0 (0x0000007fb14d0000)
        libmd4c.so.0 => /lib/aarch64-linux-gnu/libmd4c.so.0 (0x0000007fb14b0000)
        libfreetype.so.6 => /lib/aarch64-linux-gnu/libfreetype.so.6 (0x0000007fb13d0000)
        libz.so.1 => /lib/aarch64-linux-gnu/libz.so.1 (0x0000007fb1390000)
        libicui18n.so.72 => /lib/aarch64-linux-gnu/libicui18n.so.72 (0x0000007fb1050000)
        libicuuc.so.72 => /lib/aarch64-linux-gnu/libicuuc.so.72 (0x0000007fb0e30000)
        libdouble-conversion.so.3 => /lib/aarch64-linux-gnu/libdouble-conversion.so.3 (0x0000007fb0df0000)
        libb2.so.1 => /lib/aarch64-linux-gnu/libb2.so.1 (0x0000007fb0dd0000)
        libpcre2-16.so.0 => /lib/aarch64-linux-gnu/libpcre2-16.so.0 (0x0000007fb0d30000)
        libzstd.so.1 => /lib/aarch64-linux-gnu/libzstd.so.1 (0x0000007fb0c70000)
        libelf.so.1 => /lib/aarch64-linux-gnu/libelf.so.1 (0x0000007fb0c30000)
        liblzma.so.5 => /lib/aarch64-linux-gnu/liblzma.so.5 (0x0000007fb0be0000)
        libbz2.so.1.0 => /lib/aarch64-linux-gnu/libbz2.so.1.0 (0x0000007fb0bb0000)
        libGLdispatch.so.0 => /lib/aarch64-linux-gnu/libGLdispatch.so.0 (0x0000007fb0a20000)
        libexpat.so.1 => /lib/aarch64-linux-gnu/libexpat.so.1 (0x0000007fb09d0000)
        libxcb.so.1 => /lib/aarch64-linux-gnu/libxcb.so.1 (0x0000007fb0980000)
        libpcre2-8.so.0 => /lib/aarch64-linux-gnu/libpcre2-8.so.0 (0x0000007fb08d0000)
        libdbus-1.so.3 => /lib/aarch64-linux-gnu/libdbus-1.so.3 (0x0000007fb0850000)
        libgraphite2.so.3 => /lib/aarch64-linux-gnu/libgraphite2.so.3 (0x0000007fb0810000)
        libbrotlidec.so.1 => /lib/aarch64-linux-gnu/libbrotlidec.so.1 (0x0000007fb07e0000)
        libicudata.so.72 => /lib/aarch64-linux-gnu/libicudata.so.72 (0x0000007fae9f0000)
        libgomp.so.1 => /lib/aarch64-linux-gnu/libgomp.so.1 (0x0000007fae980000)
        libXau.so.6 => /lib/aarch64-linux-gnu/libXau.so.6 (0x0000007fae960000)
        libXdmcp.so.6 => /lib/aarch64-linux-gnu/libXdmcp.so.6 (0x0000007fae940000)
        libsystemd.so.0 => /lib/aarch64-linux-gnu/libsystemd.so.0 (0x0000007fae850000)
        libbrotlicommon.so.1 => /lib/aarch64-linux-gnu/libbrotlicommon.so.1 (0x0000007fae810000)
        libbsd.so.0 => /lib/aarch64-linux-gnu/libbsd.so.0 (0x0000007fae7d0000)
        libcap.so.2 => /lib/aarch64-linux-gnu/libcap.so.2 (0x0000007fae7a0000)
        libgcrypt.so.20 => /lib/aarch64-linux-gnu/libgcrypt.so.20 (0x0000007fae690000)
        liblz4.so.1 => /lib/aarch64-linux-gnu/liblz4.so.1 (0x0000007fae640000)
        libmd.so.0 => /lib/aarch64-linux-gnu/libmd.so.0 (0x0000007fae610000)
        libgpg-error.so.0 => /lib/aarch64-linux-gnu/libgpg-error.so.0 (0x0000007fae5c0000)
```

As you can see, only  `libcamera.so` and  ` libcamera-base.so` are detected as dependencies. IPA is "image processing algorithm", practically speaking there are 3 .so files and 3 .sign files:
```
ls /usr/local/lib/aarch64-linux-gnu/libcamera/
ipa_rkisp1.so  ipa_rkisp1.so.sign  ipa_rpi_pisp.so  ipa_rpi_pisp.so.sign  ipa_rpi_vc4.so  ipa_rpi_vc4.so.sign
```
If these files are missing, the application will print a warning (`No IPA found in '/usr/local/lib/aarch64-linux-gnu/libcamera`), and it will not be able to open the camera and give me images from it.
Even knowing the names and path of these files, I was unable to locate the source code that loads them. I did try!

**Reply by probonopd** [reply]
> As you can see, only libcamera.so and  libcamera-base.so are detected as dependencies.

Do these get deployed into the AppDir? (They should.)

> ls /usr/local/lib/aarch64-linux-gnu/libcamera/
> ipa_rkisp1.so  ipa_rkisp1.so.sign  ipa_rpi_pisp.so  ipa_rpi_pisp.so.sign  ipa_rpi_vc4.so  ipa_rpi_vc4.so.sign

Since these are not found by `ldd`, they will not be deployed into the AppDir.

**Reply by VioletGiraffe** [reply]
Yes, all the libraries that are reported by `ldd` are deployed. This has been my question all along: how to handle dynamic libraries that are NOT reported by `ldd`. How to make the binaries that did get deployed search for their additional modules inside the AppImage structure rather than in the global `\usr\`.

**Reply by probonopd** [reply]
> This has been my question all along: how to handle dynamic libraries that are NOT reported by ldd. 

Only libraries detected by ldd are deployed.

If you think about it: How would appimagetool be able to know about the extra libraries if they are not reported by ldd?

So, you need to make the additional libraries dependencies of your program (or one of the libraries it depends on).

This can be done, e.g., by `patchelf --add-needed LIBRARY`.

**Reply by VioletGiraffe** [reply]
Thanks, I had no idea such a thing exists!
Now I'm getting an error from `appimagetool`, why could it be? 
```
ERROR getDeps: did not find library /usr/local/lib/aarch64-linux-gnu/libcamera/ipa_rpi_vc4.so
```
`stat` shows that the file exists and is a normal file. I also know it can be copied with `cp` without problems (and without `sudo`):
```
stat /usr/local/lib/aarch64-linux-gnu/libcamera/ipa_rpi_vc4.so
  File: /usr/local/lib/aarch64-linux-gnu/libcamera/ipa_rpi_vc4.so
  Size: 11564424        Blocks: 22592      IO Block: 4096   regular file
Device: 179,2   Inode: 414166      Links: 1
Access: (0755/-rwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2023-10-27 17:43:03.272917411 +0300
Modify: 2023-10-31 13:38:34.375353010 +0200
Change: 2023-10-31 13:38:34.375353010 +0200
 Birth: 2023-10-31 13:38:34.035362315 +0200
```

**Reply by probonopd** [reply]
How exactly did you invoke `patchelf`?

What is your `ldd` output now?

**Reply by VioletGiraffe** [reply]
I invoked it like so:
```
for file in /usr/local/lib/aarch64-linux-gnu/libcamera/*.so; do
   if [ -e "$file" ]; then
       patchelf --add-needed "$file" ./AppDir/usr/bin/"$app_name"
   fi
done
```
`ldd` for my binary now prints:
```
linux-vdso.so.1 (0x0000007fb4ab7000)
/usr/local/lib/aarch64-linux-gnu/libcamera/ipa_rpi_vc4.so (0x0000007fb46f0000)
/usr/local/lib/aarch64-linux-gnu/libcamera/ipa_rpi_pisp.so (0x0000007fb44a0000)
/usr/local/lib/aarch64-linux-gnu/libcamera/ipa_rkisp1.so (0x0000007fb4370000)
libudev.so.1 => /lib/aarch64-linux-gnu/libudev.so.1 (0x0000007fb4310000)
libcamera-base.so.0.1 => /usr/local/lib/aarch64-linux-gnu/libcamera-base.so.0.1 (0x0000007fb4260000)
...
```

**Reply by probonopd** [reply]
It looks like you encoded the absolute paths. I wasn't aware that this was even allowed.

Try encoding the filenames onl without their path. Start with a clean binary (on which patchelf never ran before), and try:

```
for file in /usr/local/lib/aarch64-linux-gnu/libcamera/*.so; do
   if [ -e "$file" ]; then
       patchelf --add-needed $(dirname "$file") ./AppDir/usr/bin/"$app_name"
   fi
done
```

**Reply by VioletGiraffe** [reply]
Thank you, that works in conjunction with `export LD_LIBRARY_PATH=/usr/local/lib/aarch64-linux-gnu/libcamera/`, in the sense that AppImage is produced and contains these libraries. But when I try running in on a clean system, I still get:
```
WARN IPAManager ipa_manager.cpp:154 No IPA found in '/usr/local/lib/aarch64-linux-gnu/libcamera'
```

**Reply by probonopd** [reply]
You need to check the code that produces the `No IPA found` message. Which files is it trying to load from which exact paths, and do those files exist at these paths in the AppDir? (Must not be absolute paths inside `/usr/local/...` because those won't exist on your users' machines)

**Reply by VioletGiraffe** [reply]
I think I've found the relevant code and a way to specify a different path via an environment variable. But what about a universal solution? `AppRun` knows `/usr/local/lib/...` paths should be redirected, is there no way to handle this?

**Reply by probonopd** [reply]
`AppRun` does not "redirect" any paths. You need to change the code so that the paths to the files you want to load are calculated relative to the main binary, instead of hardcoding absolute paths.

More information:
https://docs.appimage.org/reference/best-practices.html?highlight=relocatable#binaries-must-not-use-compiled-in-absolute-paths

**Comment by VioletGiraffe** [comment]
Hi @probonopd, a quick question, if I may, which seems close enough to the original topic. How do I correctly deploy more than one executable with `go-appimagetool`? Not .so, but a top-level executable that will be called at runtime by the main executable. I can `cp` it to the AppDir, but how to tell `appimagetool` to deploy its dependencies?

**Reply by probonopd** [reply]
Not sure if I had implemented this yet. Usually every application should get its own AppImage. The main exception being auxiliary applications that are invoked by the main application. Possibly this still would need to be implemented, don't quite remember.
